scripts/gha/test_lab.py

```diff
@@ -0,0 +1,334 @@
+    log_text = _get_testapp_log_text_from_gcs(test.results_dir)
+    if log_text:
+      results = test_validation.validate_results(log_text, code_platform)
+      logging.info("Test summary or log tail:\n%s", results.summary)
+      if not results.complete:
+        testapp_errors.append(test.testapp_path)
+      if results.fails > 0:
+        testapp_failures.append(test.testapp_path)
+    else:
+      testapp_errors.append(test.testapp_path)
+
+  num_tests = len(tests)
+  num_fails = len(testapp_failures)
+  num_errors = len(testapp_errors)
+  num_successes = num_tests - num_fails - num_errors
+
+  logging.info("OVERALL TEST SUMMARY:")
+  logging.info(
+      "%d TESTAPPS QUEUED, %d SUCCESSES, %d FAILURES, %d ERRORS",
+      num_tests, num_successes, num_fails, num_errors)
+  if testapp_failures:
+    logging.info("TEST FAILURES:\n%s", "\n".join(testapp_failures))
+  if testapp_errors:
+    logging.info("TEST ERRORS:\n%s", "\n".join(testapp_errors))
+  return num_successes == num_tests
+
+
+def _get_testapp_log_text_from_gcs(results_dir):
+  """Gets the testapp log text generated by game loops."""
+  gcs_dir = _relative_path_to_gs_uri(results_dir)
+  try:
+    gcs_contents = _gcs_list_dir(gcs_dir)
+  except subprocess.CalledProcessError as e:
+    logging.error("Unexpected error searching GCS logs:\n%s", e)
+    return None
+  # The path to the testapp log depends on the platform, device, and scenario
+  # being tested. We search for a json file with 'results' in the name to avoid
+  # hard-coding too many assumptions about the path. The testapp log should be
+  # the only json, but 'results' adds some redundancy in case this changes.
+  matching_gcs_logs = [
+      line for line in gcs_contents
+      if line.endswith(".json") and "results" in line.lower()
+  ]
+  if not matching_gcs_logs:
+    logging.error("Failed to find results log on GCS.")
+    return None
+  # This assumes testapps only have one scenario. Could change in the future.
+  if len(matching_gcs_logs) > 1:
+    logging.warning("Multiple scenario logs found.")
+  gcs_log = matching_gcs_logs[0]
+  try:
+    logging.info("Found results log: %s", gcs_log)
+    log_text = _gcs_read_file(gcs_log)
+    if not log_text:
+      logging.warning("Testapp log is empty. App may have crashed.")
+    return log_text
+  except subprocess.CalledProcessError as e:
+    logging.error("Unexpected error reading GCS log:\n%s", e)
+    return None
+
+
+def _relative_path_to_gs_uri(path):
+  """Converts a relative GCS path to a GS URI understood by gsutil."""
+  # e.g. <results_dir> -> gs://<project_id>/<results_dir>
+  return "gs://%s/%s" % (_PROJECT_ID, path)
+
+
+def _gcs_list_dir(gcs_path):
+  """Recursively returns a list of contents for a directory on GCS."""
+  args = ["gsutil", "ls", "-r", gcs_path]
+  logging.info("Listing GCS contents: %s", " ".join(args))
+  result = subprocess.run(args=args, capture_output=True, text=True, check=True)
```

Comment:
We have a function to run commands in `utils.py`. Though its trivial here, it would be nice to use common functions from a maintenance point of view (unless it doesnt meet your requirements).

Reply:
From a maintenance point of view I think it's debatable in cases like this (invoking a single function with minimal additional logic): adding that bit of indirection means debugging requires tracking logic across another file and then unpacking the logic in that method. The only benefit would be if we needed to make changes to all uses of subprocess.run (if Python changed the api, for example), which wouldn't be a hard change regardless. 

As it so happens, after looking at the implementation in utils, it actually won't serve in all of my use cases since it always sets text to True if capture_output is True, which will cause the output to be encoded and normalized. This is usually fine, but not in all cases, namely when reading binary files like archives or encrypted files.

Reply:
Another benefit is that it would standardize the way we are calling commands in various python scripts. 
We could address the `text` argument by setting its default to `None` and later setting it to `captureOutput` only if caller hasn't specified anything. 

I think desktop/android/ios(possibly) will probably be sharing these functions as much as possible.  Let me know what you think and I am okay with ignoring this and coming back to it if we feel it might be useful to have it in a central place.

Reply:
I don't feel too strongly about it, but adding a dependency on a custom module just for this feels a bit icky to me. I say we kick this bucket down the road - we should do a pass when everything's migrated over to see what we can refactor and clean up anyway.

Reply:
Makes sense!